File: Assignment 4/Python/09. detect loop in a list .py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

def create_node(n):
    return Node(n)

def is_loop(head):
    slow = head
    fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
    return False

def traversal(head):
    ptr = head
    while ptr:
        print(f"{ptr.val} --> ", end="")
        ptr = ptr.next
    print("NULL")

if __name__ == "__main__":
    head = create_node(1)
    snd = create_node(2)
    trd = create_node(3)
    four = create_node(4)

    head.next = snd
    snd.next = trd
    trd.next = four
    four.next = snd  # Create a loop

    if is_loop(head):
        print("Loop found")
    else:
        print("No loop")
        traversal(head)


# OUTPUT:
# Loop found

File: Assignment 4/C/09. detect loop in a list.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct node{
    int val;
    struct node* next;
}node;

node* createNode(int n){
    node* t = (node*)malloc(sizeof(node));
    t->val = n;
    t->next = NULL;
    return t;
}

bool isLoop(node* head){
    node* slow = head;
    node* fast = head;
    while(fast && fast->next){
        slow = slow->next;
        fast = fast->next->next;
        if(slow==fast){
            return true;
        }
    }
    return false;
}

void traversal(node* head){
    node* ptr = head;
    while(ptr){
        printf("%d --> ",ptr->val);
        ptr = ptr->next;
    }
    printf("NULL\n");
}

int main(){
    node* head = createNode(1);
    node* snd = createNode(2);
    node* trd = createNode(3);
    node* four = createNode(4);

    head->next = snd;
    snd->next = trd;
    trd->next = four;
    four->next = snd;  // loop
    
    if(isLoop(head)){
        printf("loop found");
    }else{
        printf("no loop\n");
        traversal(head);
    }
    
    return 0;
}


//OUTPUT:

// loop found
